Migrate intro-to-airflow guide (#1177)
astronomer/docs: guides/intro-to-airflow.md

---
title: "An introduction to Apache Airflow"

> sidebar_label: "An introduction to Apache Airflow"

Reviewer: ```suggestion
sidebar_label: "Introduction to Airflow"
```

Author: @jwitz should we be truncating the product name _Apache Airflow_? If this is our standard should it be documented in our style guide? The Google developer documentation style guide indicates that we should always use "...the full trademarked product name".

> description: "Get started with Apache Airflow."

Reviewer: ```suggestion
description: "Learn what Airflow is and how it's used."
```

![Example Use Case](/img/guides/example_pipeline.png)

Some common use cases of Airflow include:

- ETL/ELT pipelines: For example, running a write, audit, publish pattern on data in Snowflake as shown in the example implementation in [Orchestrating Snowflake Queries with Airflow](https://www.astronomer.io/guides/airflow-snowflake/).
- MLOps: For example, using Airflow with Tensorflow and MLFlow as shown in [Using Airflow with Tensorflow and MLFlow](https://www.astronomer.io/events/webinars/using-airflow-with-tensorflow-mlflow/).
- Operationalized analytics: For example, orchestrating a pipeline to extract insights from your data and display them in dashboards, an implementation of which is described in [Using Airflow as a Data Analyst](https://www.astronomer.io/events/webinars/using-airflow-as-a-data-analyst/).

## Core Airflow concepts

To navigate Airflow resources, it is helpful to have a general understanding of the following core Airflow concepts:

> - DAG: Directed Acyclic Graph. An Airflow DAG is a workflow defined as a graph, where all dependencies between nodes are directed and nodes do not self-reference. For more information on Airflow DAGs, see [Introduction to Airflow DAGs](https://www.astronomer.io/guides/dags/).
> - DAG run: The execution of a DAG at a specific point in time. A DAG run can be scheduled or manually triggered.
> - Task: A node in a DAG graph describing one unit of work.
> - Task instance: The combination of a task, in a specific DAG, being executed at a specific point in time.

Reviewer: ```suggestion
- **DAG**: Directed Acyclic Graph. An Airflow DAG is a workflow defined as a graph, where all dependencies between nodes are directed and nodes do not self-reference. For more information on Airflow DAGs, see [Introduction to Airflow DAGs](https://www.astronomer.io/guides/dags/).
- **DAG run**: The execution of a DAG at a specific point in time. A DAG run can be scheduled or manually triggered.
- **Task**: A step in a DAG describing a single unit of work.
- **Task instance**: The execution of a task at a specific point in time.
```

When authoring DAGs, you use operators. An operator is an abstraction over Python code designed to perform a specific action and takes the form of a function that accepts parameters. Each operator in your DAG code corresponds to one Airflow task.

The following are the primary operator categories:

> - Action operators: Execute a function. For example, the [`PythonOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) or the [`BashOperator`](https://www.astronomer.io/guides/scripts-bash-operator/).
> - Transfer operators: Move data from a source to a destination. For example, the [`S3ToRedshiftOperator`](https://registry.astronomer.io/providers/amazon/modules/s3toredshiftoperator).
> - [Sensors](https://www.astronomer.io/guides/what-is-a-sensor/): Wait for something to happen. For example, the [`ExternalTaskSensor`](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor) or the [`HttpSensorAsync`](https://registry.astronomer.io/providers/astronomer-providers/modules/httpsensorasync).

Reviewer: ```suggestion
- **Action operators**: Execute a function. For example, the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) or the [BashOperator](https://www.astronomer.io/guides/scripts-bash-operator/).
- **Transfer operators**: Move data from a source to a destination. For example, the [S3ToRedshiftOperator](https://registry.astronomer.io/providers/amazon/modules/s3toredshiftoperator).
- **[Sensors](https://www.astronomer.io/guides/what-is-a-sensor/)**: Wait for something to happen. For example, the [ExternalTaskSensor](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor) or the [HttpSensorAsync](https://registry.astronomer.io/providers/astronomer-providers/modules/httpsensorasync).
```

While operators are defined individually, they can pass information to each other by using [XComs](https://www.astronomer.io/guides/airflow-passing-data-between-tasks/). To learn more about operators, see [Operators 101](https://www.astronomer.io/guides/what-is-an-operator/).

> Some commonly used action operators such as `PythonOperator` are part of core Airflow and are automatically installed in your Airflow instance. Operators used to interact with external systems are maintained separately to Airflow in provider packages.

Reviewer: ```suggestion
Some commonly used action operators such as the PythonOperator are part of core Airflow and are automatically installed in your Airflow instance. Operators used to interact with external systems are maintained separately to Airflow in provider packages.
```

Providers are packages that are maintained by the community and include all of the core operators, hooks, and sensors for a given service, for example:

- [Amazon provider](https://registry.astronomer.io/providers/amazon)
- [Snowflake provider](https://registry.astronomer.io/providers/snowflake)
- [Google provider](https://registry.astronomer.io/providers/google)
- [Azure provider](https://registry.astronomer.io/providers/microsoft-azure)
- [Databricks provider](https://registry.astronomer.io/providers/databricks)
- [Fivetran provider](https://registry.astronomer.io/providers/fivetran)

> To view the available providers and operators, see the [Astronomer Registry](https://registry.astronomer.io/). In many cases, when you need to connect Airflow to an external application you'll need an [Airflow Connection](https://www.astronomer.io/guides/connections/).

Reviewer: ```suggestion
To view the available providers and operators, see the [Astronomer Registry](https://registry.astronomer.io/).
```

## Airflow components

When working with Airflow, it is important to understand the underlying components of its infrastructure. Knowing which components are available and their purpose can help you develop your DAGs, troubleshoot issues, and run Airflow successfully.

The following Airflow components must be running at all times:

> - Webserver: A Flask server running with Gunicorn that serves the [Airflow UI](https://www.astronomer.io/guides/airflow-ui/).
> - [Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html): A Daemon responsible for scheduling jobs. This is a multi-threaded Python process that determines what tasks need to be run, when they need to be run, and where they are run.
> - [Database](https://www.astronomer.io/guides/airflow-database): A database where all DAG and task metadata are stored. This is typically a Postgres database, but MySQL, MsSQL, and SQLite are also supported.
> - [Executor](https://www.astronomer.io/guides/airflow-executors-explained/): The mechanism that defines how the available computing resources are used to execute tasks. An executor is running within the Scheduler whenever Airflow is up.

Reviewer: ```suggestion
- **Webserver**: A Flask server running with Gunicorn that serves the [Airflow UI](https://www.astronomer.io/guides/airflow-ui/).
- **[Scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html)**: A Daemon responsible for scheduling jobs. This is a multi-threaded Python process that determines what tasks need to be run, when they need to be run, and where they are run.
- **[Database](https://www.astronomer.io/guides/airflow-database)**: A database where all DAG and task metadata are stored. This is typically a Postgres database, but MySQL, MsSQL, and SQLite are also supported.
- **[Executor](https://www.astronomer.io/guides/airflow-executors-explained/)**: The mechanism that defines how the available computing resources are used to execute tasks. An executor is running within the scheduler whenever Airflow is up.
```

Additionally, you may also have the following situational components:

> - Triggerer: A separate process which supports [deferrable operators](https://www.astronomer.io/guides/deferrable-operators). This component is optional and must be run separately.
> - Worker: The process that executes tasks, as defined by the executor. Depending on which executor you choose, you may or may not have workers as part of your Airflow infrastructure.

Reviewer: ```suggestion
- **Triggerer**: A separate process which supports [deferrable operators](https://www.astronomer.io/guides/deferrable-operators). This component is optional and must be run separately.
- **Worker**: The process that executes tasks, as defined by the executor. Depending on which executor you choose, you may or may not have workers as part of your Airflow infrastructure.
```
